src/core_plugins/vega/public/vega_view/vega_base_view.js

```diff
@@ -0,0 +1,160 @@
+
+      // bypass the onWarn warning checks - in some cases warnings may still need to be shown despite being disabled
+      for (const warn of this._parser.warnings) {
+        this._addMessage('warn', warn);
+      }
+
+      if (this._parser.error) {
+        this._addMessage('err', this._parser.error);
+        return;
+      }
+
+      this._$container = $('<div class="vega-view-container">')
+        .appendTo(this._$parentEl);
+      this._$controls = $('<div class="vega-controls-container">')
+        .css('flex-direction', this._parser.controlsDir)
+        .appendTo(this._$parentEl);
+
+      this._addDestroyHandler(() => {
+        this._$container.remove();
+        this._$container = null;
+        this._$controls.remove();
+        this._$controls = null;
+        if (this._$messages) {
+          this._$messages.remove();
+          this._$messages = null;
+        }
+      });
+
+      this._vegaViewConfig = {
+        logLevel: vega.Warn,
+        renderer: this._parser.renderer,
+      };
+      if (!this._vegaConfig.enableExternalUrls) {
+        // Override URL loader to disable all URL-based requests
+        const loader = vega.loader();
+        loader.load = () => {
+          throw new Error('External URLs have been disabled in kibana.yml');
+        };
+        this._vegaViewConfig.loader = loader;
+      }
+
+      // The derived class should create this method
+      await this._initViewCustomizations();
+    } catch (err) {
+      this.onError(err);
+    }
+  }
+
+  onError() {
+    this._addMessage('err', Utils.formatErrorToStr(...arguments));
+  }
+
+  onWarn() {
+    if (!this._parser || !this._parser.hideWarnings) {
+      this._addMessage('warn', Utils.formatWarningToStr(...arguments));
+    }
+  }
+
+  _addMessage(type, text) {
+    if (!this._$messages) {
+      this._$messages = $(`<ul class="vega-messages">`).appendTo(this._$parentEl);
+    }
+    this._$messages.append(
+      $(`<li class="vega-message-${type}">`).append(
+        $(`<pre>`).text(text)
+      )
+    );
+  }
+
+  resize() {
+    if (this._parser.useResize && this._view && this.updateVegaSize(this._view)) {
+      return this._view.runAsync();
+    }
+  }
+
+  updateVegaSize(view) {
+    // For some reason the object is slightly scrollable without the extra padding.
+    // This might be due to https://github.com/jquery/jquery/issues/3808
+    // Which is being fixed as part of jQuery 3.3.0
+    const heightExtraPadding = 6;
+    const width = Math.max(0, this._$container.width() - this._parser.paddingWidth);
+    const height = Math.max(0, this._$container.height() - this._parser.paddingHeight) - heightExtraPadding;
+    if (view.width() !== width || view.height() !== height) {
+      view.width(width).height(height);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Set global debug variable to simplify vega debugging in console. Show info message first time
+   */
+  static setDebugValues(view, spec, vlspec) {
```

Comment:
I thought, that there was already an discussion ongoing (but don't see any comments anymore about this).

I don't think we should log anything out to console whenever there is a vega visualization and attach something to the `window` object. There can be multiple vega vis visible (e.g. when placed on a dashboard). This code would give access to ("randomly") one of it.

If you want to give access to some of this, please write a spy panel for this, that only shows for Vega visualizations.
This could show the spec and view in serialized forms. Since I assume this is rather another large step, I think we could safely add that debug panel at a later time, but I would still like, not to have this debug output in there.

Reply:
I agree that we should move this to the spy panel. Would you be ok if this is only shown with the editor? (as this is something most users are already familiar with in the plugin and in the other vega tools).  If so, is there an easy way to tell for the Vis if its being used from the editor panel or from its own?  I know it breaks the ideal compartmental model, and I will fix that with the elaborate spy panel in the future.

Reply:
Funny enough, react does the same thing in Kibana:  **Download the React DevTools for a better development experience: https://fb.me/react-devtools** is shown in the dev console ;)
In any case, this is done: https://github.com/elastic/kibana/pull/15014/commits/28324c5f9f69d79d221ac79171c6810b51d9cf38